src/jacobi.c

```diff
@@ -0,0 +1,219 @@
+        if (i == cIndex1 || i == cIndex2) {  
+            output[i] = c;
+        } else if (i == sIndex1) {
+            output[i] = s;
+        } else if (i == sIndex2) {
+            output[i] = -s;
+        } else if (i / size == i % size) {
+            output[i] = 1; // Diagonal
+        } else {
+            output[i] = 0; // Off diagonal
+        }
+    }
+}
+
+
+/**
+ * Creates the identity matrix (1 on the diagonal and 0 on other coordinates)
+ * @param size size of the matrix
+ * @param output output matrix
+*/
+void jacobiCreateIdentityMatrix(vec_size size, real_number* output) {
+    for (vec_size i = 0; i < size * size; ++i) {
+        if (i / size == i % size) {
+            output[i] = 1;
+        } else {
+            output[i] = 0;
+        }
+    }
+}
+
+/**
+ * Computes the off diagonal sum of a matrix (sums every element not in the diagonal)
+ * @param matrix matrix to compute the sum on
+ * @param size size of matrix
+*/
+vec_size jacobiComputeOffDiagonalSum(real_number* matrix, vec_size size) {
+    vec_size sum = 0;
+    
+    for (vec_size i = 0; i < size * size; ++i) {
+        if (i / size != i % size) { // Off diagonal
+            sum += matrix[i];
+        }
+    }
+    return sum;
+}
+
+/**
+ * Computes the jacobi method to find the eigenvalues and eigenvectors of the input matrix
+ * @param inputMatrix matrix whose eigenvalues and eigenvectors we are looking for (will contain the eigenvalues after this function has been called)
+ * @param size size of the matrix (matrix must be square)
+ * @param outputMatrix matrix containing every eigenvector of the input matrix (same size as the input matrix)
+ * @param iterations number of iterations of the algorithm
+ * @param cyclic whether to use the cyclic jacobi method or the classical version
+*/
+void jacobi(real_number* inputMatrix, vec_size size, real_number* outputMatrix, vec_size iterations, vec_size cyclic) {
+
+    vec_size sizes3d[3] = {size, size, size};
+    vec_size sweepSize = size * (size - 1) / 2; // A sweep is defined as a n * (n - 1) / 2 jacobi rotations
+    real_number p[size * size]; // Rotation matrix
+    real_number vBuffer[size * size];
+    real_number aBuffer[size * size];
+    vec_size maxRow, maxCol, indexToMinimize;
+    vec_size sizeSquared = size * size;
+
+    jacobiCreateIdentityMatrix(size, outputMatrix);
+
+    for (vec_size i = 0; i < iterations; ++i) {
+
+        if (cyclic) {
+            if (indexToMinimize == (sizeSquared - 2)) {
+                indexToMinimize = 0;
+            }
+
+            indexToMinimize = (indexToMinimize + 1) % (sizeSquared);
+            if (indexToMinimize / size == indexToMinimize % size) {
+                indexToMinimize = (indexToMinimize + 1) % (sizeSquared);
+            }
+
+            maxRow = indexToMinimize / size;
+            maxCol = indexToMinimize % size;
+        } else {
+            jacobiMaxIndex(inputMatrix, size, &maxRow, &maxCol);
+        }
+
+        if (i < 3 * sweepSize) {
```

Comment:
magic number, 3?

Reply:
This is an optimization proposed in Numerical Recipes

Reply:
Ok, do we know why we are using 3 here? I think we should either add a comment explaining this number or replace 3 with a constant equal to 3

Reply:
I have a comment right under the condition, but I can add the fact that this number was proposed in the book Numerical Recipes

Reply:
The comment under it explains what we do during the first 3 sweeps, but not why we chose that number. If we don't know precisely why we decided to use 3 we should add a comment explaining that it was proposed in the book.